src/ray/core_worker/task_manager.h

```diff
@@ -87,6 +87,51 @@ using PushErrorCallback = std::function<Status(const JobID &job_id,
                                                const std::string &error_message,
                                                double timestamp)>;
 
+/// When the streaming generator tasks are submitted,
+/// the intermediate return objects are streamed
+/// back to the task manager.
+/// This class manages the references of intermediately
+/// streamed object references.
+/// The API is not thread-safe.
+class ObjectRefStream {
+ public:
+  ObjectRefStream(const ObjectID &generator_id) : generator_id_(generator_id) {}
+
+  /// Asynchronously read object reference of the next index.
+  ///
+  /// \param[out] object_id_out The next object ID from the stream.
+  /// Nil ID is returned if the next index hasn't been written.
+  /// \return KeyError if it reaches to EoF. Ok otherwise.
+  Status AsyncReadNext(ObjectID *object_id_out);
+
+  /// Write the object id to the stream of an index idx.
+  ///
+  /// If the idx has been already read (by AsyncReadNext),
+  /// the write request will be ignored. If the idx has been
+  /// already written, it will be no-op. It doesn't override.
+  ///
+  /// \param[in] object_id The object id that will be read at index idx.
+  /// \param[in] idx The index where the object id will be written.
+  /// \return True if the idx hasn't been used. False otherwise.
+  bool Write(const ObjectID &object_id, int64_t idx);
```

Comment:
AppendToStream be a better name?

Reply:
Actually, I will use InsertToStream in this case since Append means it appends to the end (but this API allows to specify an index).